Add feature flag for profile admin section

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adds a feature flag `:profile_admin` which removes the menu item from the admin index and sidebar and also disables the routes. 

Anyone should feel free to merge this, it has tests for all scenarios.

## Related Tickets & Documents

https://github.com/forem/InternalProjectPlanning/issues/147

## QA Instructions, Screenshots, Recordings

### Admin index:

0. Ensure the feature flag is disabled.
1. Go to `/admin`
2. Ensure that you don't see the profile admin link (it would be right after "Config")
    <img width="1289" alt="Screen Shot 2020-10-29 at 10 15 24" src="https://user-images.githubusercontent.com/47985/97521218-ac070d00-19cf-11eb-8314-ec6c56f648e3.png">
3. Enable the feature flag in the Rails console:
    ```ruby
    Flipper.enable(:profile_admin)
    ```
4. Reload `/admin`, you should now see a "Config: Profile Setup" option:
    <img width="1270" alt="Admin_Portals" src="https://user-images.githubusercontent.com/47985/97521336-eec8e500-19cf-11eb-9adb-4b9302ee277a.png">


### Admin sidebar menu:

0. Ensure the feature flag is disabled.
1. Go an admin page, e.g. `/admin/articles`
2. Ensure that you don't see the profile admin link (it would be right under "Config")
    <img width="319" alt="Screen Shot 2020-10-30 at 11 31 52" src="https://user-images.githubusercontent.com/47985/97660234-851a0b00-1aa3-11eb-9aaf-1c4d26d2eec1.png">
3. Enable the feature flag in the Rails console:
    ```ruby
    Flipper.enable(:profile_admin)
    ```
4. Reload the page, you should now see a "Config: Profile Setup" option:
    <img width="293" alt="Articles" src="https://user-images.githubusercontent.com/47985/97660316-bd214e00-1aa3-11eb-8ff0-8aa932932244.png">

### Routing constraint

0. Ensure the feature flag is disabled.
1. Go to `/admin/profile_fields`
2. You should be redirected to our root route.
3. Enable the feature flag in the Rails console:
    ```ruby
    Flipper.enable(:profile_admin)
    ```
4. Going to `/admin/profile_fields` should now render the profile admin page.

## Added tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] Docs.forem.com
- [ ] README
- [X] No documentation needed

citizen428 commented on app/views/admin/admin_portals/index.html.erb:
```diff
@@ -1,5 +1,12 @@
   <div class="grid gap-2 m:gap-4 l:gap-6 m:grid-cols-2 l:grid-cols-3 px-2 m:px-0">
-    <% Admin::ApplicationController::MENU_ITEMS.each do |menu_item| %>
+    <%# Temporary code while we use a feature flag for the profile admin. %>
```
This is not pretty. BUT: dealing with frozen constants is a bit painful and I didn't feel like adding a whole new helper for something that's hopefully quite temporary. If people prefer that I can extract it though.

jacobherrington reviewed (approved):
Works for me :+1:

vaidehijoshi reviewed (changes requested):
Yay, feature flags!! 👏 

Can we remove it from the sidebar when the feature is disabled as well? I wonder if we could just disable the entire `/admin/profile_fields` route and hide _that_ behind a feature flag (in my opinion, that would be the ideal solution here so that we don't need to worry about every place that we link to it), but I think we should definitely remove it [from the sidebar](https://github.com/forem/forem/blob/5457f0c0b702af71cc864a0fbdd065238760b7f5/app/controllers/admin/application_controller.rb#L29) since that will show up on every admin view :) 

<img width="1215" alt="Screen_Shot_2020-10-29_at_9_27_58_AM" src="https://user-images.githubusercontent.com/6921610/97603091-50398580-19c9-11eb-826c-870c293b91ef.png">

citizen428 commented:
@vaidehijoshi 

> Can we remove it from the sidebar when the feature is disabled as well? 

Definitely, my oversight, thanks for catching that.

>  wonder if we could just disable the entire `/admin/profile_fields` route and hide that behind a feature flag (in my opinion, that would be the ideal solution here so that we don't need to worry about every place that we link to it)

Sure, we can disable the route, but we'll still have to remove the links too, otherwise they lead to a 404.

citizen428 commented:
@vaidehijoshi I now did the following:

1. Add a new `AdminHelper` module with an `admin_menu_items` method that is used for both the sidebar partial and admin index view with keeps all the logic in one place.
2. Added a spec for the sidebar too.
3. Disabled all admin profile routes via a constraint and added a routing spec for that. The spec only tests for the `GET` route, but IMHO that's good enough since it's all done via a `resources` block, so if one is disabled, all are.

Re-requesting review now.

vaidehijoshi reviewed (commented):
This is great, @citizen428! Thanks for taking my feedback into account :)

I think there are just a few specs (in `profile_field_groups_spec.rb`) that need the feature flag to mocked to be on, and then this is should be good to go !

citizen428 commented:
Since it took less than a minute I fixed it now.

citizen428 commented:
Thanks @vaidehijoshi 😃 It was quick enough so I fixed it right away.

vaidehijoshi reviewed (approved):
🎉🎉🎉
